In this project we will clean and analyze data on 50,000 used cars listed in eBay Kleinanzeigen.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

There is data on 50,000 autos. Some indices that are missing values are: vehicle type, gearbox, model, fuel type, and not repaired damage. 

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos_c = autos.copy()

In [7]:
autos_c

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [8]:
autos_c.rename(columns={"yearOfRegistration":"registration_year", "monthOfRegistration":"registration_month",
                        "notRepairedDamage":"unrepaired_damage", "dateCreated":"ad_created"}, inplace=True)

In [9]:
def clean_column(col):
    res = [col[0].lower()]
    for chr in col[1:]:
        if chr in ('ABCDEFGHIJKLMNOPQRSTUVWXYZ'):
            res.append('_')
            res.append(chr.lower())
        else:
            res.append(chr)
    return ''.join(res)

new_col = []
for c in autos_c.columns:
    cleaned_auto_c = clean_column(c)
    new_col.append(cleaned_auto_c)
autos_c.columns = new_col
print(autos_c.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [10]:
autos_c.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We made the column titles lower case and snakecase. This makes the date easier to analyze and read.

In [11]:
autos_c.describe()

registration_year     power_p_s  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

nrofpictures has mostly 0s. The postal code, powerps, registration year (min of 1000 and max of 9999 do not make sense) column needs more investigation. Odometer reading (#km) and price (#$) are numerical data stored as text.

In [12]:
autos_c["price"] = autos_c["price"].str.replace("$","").str.replace(",","").astype(int)
autos_c["odometer"] = autos_c["odometer"].str.replace("km","").str.replace(",","").astype(int)

autos_c.rename({"price":"price_USD", "odometer":"odometer_km"}, axis=1, inplace=True)

autos_c

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In [13]:
odometer = autos_c["odometer_km"]
odometer.unique().shape

(13,)

There are 13 unique values for the odometer readings.

In [14]:
odometer.describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

The values for 50%, 75%, and max are all 150000. This is suspicious and will need to be explored further. The minimum reading for a car listing was 5000. If the mean was 125732, this means most of the cars had high odometer readings, closer to 150000 than 100000.

In [15]:
odometer.value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

As expected 150000 was the highest odometer reading. Because the counts at 150000 are so much more than other readings, this forces the lower values to have less significance. There are no obvious outliers. Let's look at null values to explore the odometer readings further.

In [16]:
print(odometer.value_counts(dropna=False))

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


We can see above that there are no null odometer values. Now let us explore the auto prices.

In [17]:
price = autos_c["price_USD"] 
price.unique().shape

(2357,)

There are 2357 unique prices for the autos.

In [18]:
price.describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_USD, dtype: float64

The minimum price was 0 and the maximum was 100 million. 100 million is too high and 0 is getting a car for free.

In [19]:
price.value_counts().sort_index(ascending=False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price_USD, dtype: int64

We saw that the describe() method rounded the car price (99999999) to 100 million. We will remove the 100 million auto price, as this is too high a price for a car and 0, as this is getting a car for free.

In [20]:
autos_c = autos_c[autos_c["price_USD"].between(1,27322222)]

In [21]:
autos_c.describe()

price_USD  registration_year     power_p_s    odometer_km  \
count  4.857800e+04       48578.000000  48578.000000   48578.000000   
mean   8.069542e+03        2004.753119    117.209869  125765.675820   
std    1.800356e+05          88.632571    200.652070   39793.193823   
min    1.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.200000e+03        1999.000000     71.000000  125000.000000   
50%    3.000000e+03        2004.000000    107.000000  150000.000000   
75%    7.490000e+03        2008.000000    150.000000  150000.000000   
max    2.732222e+07        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        48578.000000         48578.0  48578.000000  
mean             5.781794             0.0  50976.081230  
std              3.685768             0.0  25749.448131  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30657.000000  
50%              6.000000             0.0  49716.000000  
75%              9.000000             0.0  71665.000000  
max             12.000000             0.0  99998.000000

We know have a minimum price of 1 and a maximum of 27 million.

Now we'll analyze the date columns.

In [22]:
print(autos_c[['date_crawled', 'ad_created', 'last_seen']][0:5])

          date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50


In [23]:
print(autos_c['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
5        2016-03-21
6        2016-03-20
7        2016-03-16
8        2016-03-22
9        2016-03-16
10       2016-03-15
11       2016-03-16
12       2016-03-31
13       2016-03-23
14       2016-03-23
15       2016-04-01
16       2016-03-16
17       2016-03-29
18       2016-03-26
19       2016-03-17
20       2016-03-05
21       2016-03-06
22       2016-03-28
23       2016-03-10
24       2016-04-03
25       2016-03-21
26       2016-04-03
28       2016-03-19
29       2016-04-02
30       2016-03-14
            ...    
49968    2016-04-01
49969    2016-03-17
49970    2016-03-21
49971    2016-03-29
49972    2016-03-26
49973    2016-03-27
49975    2016-03-27
49976    2016-03-19
49977    2016-03-31
49978    2016-04-04
49979    2016-03-20
49980    2016-03-12
49981    2016-03-15
49982    2016-03-29
49983    2016-03-06
49985    2016-04-02
49986    2016-04-04
49987    2016-03-22
49988    2016-03-28


The dates for date crawled are shown above. Below is the percent distribution of the dates.

In [24]:
print(autos_c['date_crawled'].str[:10].value_counts(normalize=True,dropna=False))

2016-04-03    0.038598
2016-03-20    0.037877
2016-03-21    0.037404
2016-03-12    0.036930
2016-03-14    0.036539
2016-04-04    0.036519
2016-03-07    0.036004
2016-04-02    0.035469
2016-03-28    0.034851
2016-03-19    0.034769
2016-03-15    0.034275
2016-03-29    0.034131
2016-04-01    0.033678
2016-03-30    0.033678
2016-03-08    0.033328
2016-03-09    0.033101
2016-03-22    0.032998
2016-03-11    0.032566
2016-03-23    0.032216
2016-03-26    0.032196
2016-03-10    0.032175
2016-03-31    0.031846
2016-03-17    0.031640
2016-03-25    0.031599
2016-03-27    0.031084
2016-03-16    0.029602
2016-03-24    0.029334
2016-03-05    0.025320
2016-03-13    0.015666
2016-03-06    0.014039
2016-04-05    0.013092
2016-03-18    0.012907
2016-04-06    0.003170
2016-04-07    0.001400
Name: date_crawled, dtype: float64


We can sort this by earliest to latest dates.

In [25]:
print(autos_c['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True))

2016-03-05    0.025320
2016-03-06    0.014039
2016-03-07    0.036004
2016-03-08    0.033328
2016-03-09    0.033101
2016-03-10    0.032175
2016-03-11    0.032566
2016-03-12    0.036930
2016-03-13    0.015666
2016-03-14    0.036539
2016-03-15    0.034275
2016-03-16    0.029602
2016-03-17    0.031640
2016-03-18    0.012907
2016-03-19    0.034769
2016-03-20    0.037877
2016-03-21    0.037404
2016-03-22    0.032998
2016-03-23    0.032216
2016-03-24    0.029334
2016-03-25    0.031599
2016-03-26    0.032196
2016-03-27    0.031084
2016-03-28    0.034851
2016-03-29    0.034131
2016-03-30    0.033678
2016-03-31    0.031846
2016-04-01    0.033678
2016-04-02    0.035469
2016-04-03    0.038598
2016-04-04    0.036519
2016-04-05    0.013092
2016-04-06    0.003170
2016-04-07    0.001400
Name: date_crawled, dtype: float64


We see the data was first crawled on 03/05/2016 and last crawled on 04/07/2016. So this data represents approximately a month of crawling on the German eBay auto site. Also the distribution ranges from 0.00% to approximately 0.04%.

Now let us look at the dates the ads were created.

In [26]:
print(autos_c['ad_created'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
5        2016-03-21
6        2016-03-20
7        2016-03-16
8        2016-03-22
9        2016-03-16
10       2016-03-14
11       2016-03-16
12       2016-03-31
13       2016-03-23
14       2016-03-23
15       2016-04-01
16       2016-03-16
17       2016-03-29
18       2016-03-26
19       2016-03-17
20       2016-03-05
21       2016-03-06
22       2016-03-28
23       2016-03-10
24       2016-04-03
25       2016-03-21
26       2016-04-03
28       2016-03-19
29       2016-04-02
30       2016-03-14
            ...    
49968    2016-04-01
49969    2016-03-17
49970    2016-03-21
49971    2016-03-29
49972    2016-03-26
49973    2016-03-27
49975    2016-03-27
49976    2016-03-19
49977    2016-03-31
49978    2016-04-04
49979    2016-03-20
49980    2016-03-12
49981    2016-03-15
49982    2016-03-29
49983    2016-03-06
49985    2016-04-02
49986    2016-04-04
49987    2016-03-22
49988    2016-03-28


The percent distrribution for the dates the ads were created is shown below.

In [27]:
print(autos_c['ad_created'].value_counts(normalize=True, dropna=False))

2016-04-03 00:00:00    0.038845
2016-03-20 00:00:00    0.037939
2016-03-21 00:00:00    0.037610
2016-04-04 00:00:00    0.036889
2016-03-12 00:00:00    0.036766
2016-03-14 00:00:00    0.035181
2016-04-02 00:00:00    0.035139
2016-03-28 00:00:00    0.034975
2016-03-07 00:00:00    0.034728
2016-03-29 00:00:00    0.034069
2016-03-15 00:00:00    0.034007
2016-04-01 00:00:00    0.033678
2016-03-19 00:00:00    0.033678
2016-03-30 00:00:00    0.033493
2016-03-08 00:00:00    0.033348
2016-03-09 00:00:00    0.033163
2016-03-11 00:00:00    0.032896
2016-03-22 00:00:00    0.032813
2016-03-26 00:00:00    0.032257
2016-03-23 00:00:00    0.032052
2016-03-10 00:00:00    0.031887
2016-03-31 00:00:00    0.031887
2016-03-25 00:00:00    0.031743
2016-03-17 00:00:00    0.031290
2016-03-27 00:00:00    0.030981
2016-03-16 00:00:00    0.030117
2016-03-24 00:00:00    0.029273
2016-03-05 00:00:00    0.022891
2016-03-13 00:00:00    0.017004
2016-03-06 00:00:00    0.015316
                         ...   
2016-02-

Looking at the ad created dates in ascending order we get.

In [28]:
print(autos_c['ad_created'].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

2015-06-11 00:00:00    0.000021
2015-08-10 00:00:00    0.000021
2015-09-09 00:00:00    0.000021
2015-11-10 00:00:00    0.000021
2015-12-05 00:00:00    0.000021
2015-12-30 00:00:00    0.000021
2016-01-03 00:00:00    0.000021
2016-01-07 00:00:00    0.000021
2016-01-10 00:00:00    0.000041
2016-01-13 00:00:00    0.000021
2016-01-14 00:00:00    0.000021
2016-01-16 00:00:00    0.000021
2016-01-22 00:00:00    0.000021
2016-01-27 00:00:00    0.000062
2016-01-29 00:00:00    0.000021
2016-02-01 00:00:00    0.000021
2016-02-02 00:00:00    0.000041
2016-02-05 00:00:00    0.000041
2016-02-07 00:00:00    0.000021
2016-02-08 00:00:00    0.000021
2016-02-09 00:00:00    0.000021
2016-02-11 00:00:00    0.000021
2016-02-12 00:00:00    0.000041
2016-02-14 00:00:00    0.000041
2016-02-16 00:00:00    0.000021
2016-02-17 00:00:00    0.000021
2016-02-18 00:00:00    0.000041
2016-02-19 00:00:00    0.000062
2016-02-20 00:00:00    0.000041
2016-02-21 00:00:00    0.000062
                         ...   
2016-03-

The ads were created between 06/11/2015 and 04/07/2016. Now let us seen when they were last seen by the crawler.

In [29]:
print(autos_c['last_seen'].str[:10])

0        2016-04-06
1        2016-04-06
2        2016-04-06
3        2016-03-15
4        2016-04-01
5        2016-04-06
6        2016-03-23
7        2016-04-07
8        2016-03-26
9        2016-04-06
10       2016-04-07
11       2016-03-16
12       2016-04-06
13       2016-04-01
14       2016-04-01
15       2016-04-02
16       2016-03-18
17       2016-04-05
18       2016-04-02
19       2016-03-17
20       2016-03-07
21       2016-04-05
22       2016-04-01
23       2016-03-17
24       2016-04-07
25       2016-04-01
26       2016-04-05
28       2016-04-07
29       2016-04-02
30       2016-03-14
            ...    
49968    2016-04-05
49969    2016-03-25
49970    2016-04-07
49971    2016-03-29
49972    2016-03-28
49973    2016-03-27
49975    2016-04-05
49976    2016-04-07
49977    2016-04-06
49978    2016-04-06
49979    2016-04-07
49980    2016-04-06
49981    2016-04-05
49982    2016-03-29
49983    2016-03-09
49985    2016-04-06
49986    2016-04-06
49987    2016-03-26
49988    2016-04-07


If we generate the percent distribution for the last seen date, added by the crawler, we get the following.

In [30]:
print(autos_c["last_seen"].str[:10].value_counts(normalize=True,dropna=False))

2016-04-06    0.221808
2016-04-07    0.131912
2016-04-05    0.124768
2016-03-17    0.028079
2016-04-03    0.025197
2016-04-02    0.024929
2016-03-30    0.024764
2016-04-04    0.024476
2016-03-31    0.023797
2016-03-12    0.023797
2016-04-01    0.022788
2016-03-29    0.022356
2016-03-22    0.021368
2016-03-28    0.020874
2016-03-20    0.020647
2016-03-21    0.020647
2016-03-24    0.019762
2016-03-25    0.019206
2016-03-23    0.018527
2016-03-26    0.016798
2016-03-16    0.016448
2016-03-15    0.015871
2016-03-19    0.015830
2016-03-27    0.015645
2016-03-14    0.012598
2016-03-11    0.012372
2016-03-10    0.010663
2016-03-09    0.009613
2016-03-13    0.008893
2016-03-08    0.007431
2016-03-18    0.007349
2016-03-07    0.005393
2016-03-06    0.004323
2016-03-05    0.001070
Name: last_seen, dtype: float64


The last seen date percent distribution, ordered by date, is shown below.

In [31]:
print(autos_c["last_seen"].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True))

2016-03-05    0.001070
2016-03-06    0.004323
2016-03-07    0.005393
2016-03-08    0.007431
2016-03-09    0.009613
2016-03-10    0.010663
2016-03-11    0.012372
2016-03-12    0.023797
2016-03-13    0.008893
2016-03-14    0.012598
2016-03-15    0.015871
2016-03-16    0.016448
2016-03-17    0.028079
2016-03-18    0.007349
2016-03-19    0.015830
2016-03-20    0.020647
2016-03-21    0.020647
2016-03-22    0.021368
2016-03-23    0.018527
2016-03-24    0.019762
2016-03-25    0.019206
2016-03-26    0.016798
2016-03-27    0.015645
2016-03-28    0.020874
2016-03-29    0.022356
2016-03-30    0.024764
2016-03-31    0.023797
2016-04-01    0.022788
2016-04-02    0.024929
2016-04-03    0.025197
2016-04-04    0.024476
2016-04-05    0.124768
2016-04-06    0.221808
2016-04-07    0.131912
Name: last_seen, dtype: float64


The ads were last seen by the crawler between 03/05/2016 and 04/07/2016. The persecent distribution ranges from 0.0% to 0.2%. 

Now let us look at the distribution of the registration year.

In [32]:
autos_c["registration_year"].describe()

count    48578.000000
mean      2004.753119
std         88.632571
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The median registration year was 2004. The minimum was 1000 (before cars were invented!)  and the maximum year was 9999 (in the future!). Clearly the registration year data needs cleaning. Also, of the 50,000 autos analyzed, we have registration years for 48578 autos, meaning some values are missing.

In [33]:
autos_c["registration_year"].value_counts().sort_index(ascending=True)

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      24
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2637
2002    2486
2003    2699
2004    2703
2005    2936
2006    2671
2007    2277
2008    2215
2009    2086
2010    1589
2011    1623
2012    1310
2013     803
2014     664
2015     392
2016    1220
2017    1393
2018     471
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

The earliest logical registration year listed is 1910. Our data was only collected up to 2016 so we will use this as the maximum year.

In [34]:
autos_c = autos_c[autos_c["registration_year"].between(1910,2016)]

In [35]:
autos_c.describe()

price_USD  registration_year     power_p_s    odometer_km  \
count  4.669200e+04       46692.000000  46692.000000   46692.000000   
mean   7.717820e+03        2002.908250    117.906943  125581.255890   
std    1.649244e+05           7.190078    184.934094   39857.756314   
min    1.000000e+00        1910.000000      0.000000    5000.000000   
25%    1.250000e+03        1999.000000     75.000000  100000.000000   
50%    3.100000e+03        2003.000000    109.000000  150000.000000   
75%    7.500000e+03        2008.000000    150.000000  150000.000000   
max    2.732222e+07        2016.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        46692.000000         46692.0  46692.000000  
mean             5.826758             0.0  51097.568770  
std              3.670470             0.0  25757.115964  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30827.000000  
50%              6.000000             0.0  49828.000000  
75%              9.000000             0.0  71732.750000  
max             12.000000             0.0  99998.000000

The statistics for the registration year now show a minimum year of 1910 and a maximum of 2016. The average registration year is 2002 and the median is 2003. Now let's look at the perecentage distribution.

In [36]:
print(autos_c["registration_year"].value_counts(normalize= True).sort_index(ascending=True))

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
1938    0.000021
1939    0.000021
1941    0.000043
1943    0.000021
1948    0.000021
1950    0.000064
1951    0.000043
1952    0.000021
1953    0.000021
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000514
1961    0.000129
1962    0.000086
1963    0.000171
1964    0.000257
1965    0.000364
1966    0.000471
1967    0.000557
1968    0.000557
          ...   
1987    0.001542
1988    0.002891
1989    0.003727
1990    0.007432
1991    0.007260
1992    0.007946
1993    0.009102
1994    0.013471
1995    0.026300
1996    0.029405
1997    0.041784
1998    0.050608
1999    0.062066
2000    0.067592
2001    0.056476
2002    0.053243
2003    0.057804
2004    0.057890
2005    0.062880
2006    0.057205
2007    0.048766
2008    0.047439
2009    0.044676
2010    0.034032
2011    0.034760
2012    0.028056
2013    0.017198
2014    0.0142

Because in general the registration has increased over time and in 1987 it was only 0.0015%; we can assume that registrations before 1987 were minimal and can be ignored.Let us look at our registration year data again removing these years.

In [37]:
autos_c = autos_c[autos_c["registration_year"].between(1987,2016)]

In [38]:
autos_c.describe()

price_USD  registration_year     power_p_s    odometer_km  \
count  4.577200e+04       45772.000000  45772.000000   45772.000000   
mean   7.164805e+03        2003.461461    118.521913  126146.006292   
std    1.510737e+05           5.909554    186.268102   39368.238902   
min    1.000000e+00        1987.000000      0.000000    5000.000000   
25%    1.200000e+03        1999.000000     75.000000  125000.000000   
50%    3.000000e+03        2003.000000    109.000000  150000.000000   
75%    7.499000e+03        2008.000000    150.000000  150000.000000   
max    2.732222e+07        2016.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        45772.000000         45772.0  45772.000000  
mean             5.852246             0.0  51093.637639  
std              3.670115             0.0  25740.071395  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30851.000000  
50%              6.000000             0.0  49811.000000  
75%              9.000000             0.0  71726.000000  
max             12.000000             0.0  99998.000000

The minimum registration year is now 1987 and the maximum is 2016, while the mean is 2003.

In [39]:
print(autos_c["vehicle_type"].value_counts())

limousine     12315
kleinwagen    10517
kombi          8903
bus            3985
cabrio         2881
coupe          2289
suv            1921
andere          361
Name: vehicle_type, dtype: int64


In [40]:
print(autos_c["brand"].value_counts())

volkswagen        9682
bmw               5090
opel              4948
mercedes_benz     4329
audi              4022
ford              3194
renault           2191
peugeot           1392
fiat              1166
seat               852
skoda              761
nissan             710
mazda              707
smart              661
citroen            644
toyota             586
hyundai            468
volvo              409
mini               407
mitsubishi         383
honda              361
sonstige_autos     335
kia                330
alfa_romeo         293
suzuki             276
porsche            256
chevrolet          237
chrysler           161
dacia              123
daihatsu           116
jeep               103
subaru             100
land_rover          94
saab                76
daewoo              70
jaguar              70
rover               62
lancia              49
trabant             35
lada                23
Name: brand, dtype: int64


Let us determine brands that represent more than 5% of the vehicles.

In [41]:
brands = autos_c["brand"].value_counts(normalize=True) # assigns
#the brand percentage value counts to variable named brands
top_brands = brands[brands>0.05].index #assigns brands with >5%
#value counts to variable named top_brands
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


Now we will determine the mean price of each of these top 5% vehicles.

In [42]:
brand_mean_price = {}

for brand in top_brands:
    selected_brand = autos_c[autos_c["brand"] == brand]
    price_mean = selected_brand['price_USD'].mean()
    brand_mean_price[brand] = int(price_mean)
print(brand_mean_price)  

{'bmw': 8577, 'volkswagen': 6516, 'mercedes_benz': 8495, 'ford': 3751, 'audi': 9363, 'opel': 5447}


In [64]:
brand_mean_price = {}
brand_mean_mileage = {}

for brand in top_brands:
    selected_brand = autos_c[autos_c["brand"] == brand]
    price_mean = selected_brand['price_USD'].mean()
    mileage_mean = selected_brand['odometer_km'].mean()
    brand_mean_price[brand] = int(price_mean)
    brand_mean_mileage[brand] = int(mileage_mean)
    bmp_series = pd.Series(brand_mean_price)
    bmm_series = pd.Series(brand_mean_mileage)
    df = pd.DataFrame(bmp_series, columns=["Mean_Price"])
    df.insert(1, "Mean_Mileage", bmm_series)
print("Brand Mean Price", brand_mean_price)  
print("Brand Mean Mileage", brand_mean_mileage)
print("Brand Mean Price Series Constructor", bmp_series)
print("Brand Mean Mileage Series Constructor", bmm_series)
print(df)

Brand Mean Price {'bmw': 8577, 'volkswagen': 6516, 'mercedes_benz': 8495, 'ford': 3751, 'audi': 9363, 'opel': 5447}
Brand Mean Mileage {'bmw': 132743, 'volkswagen': 129153, 'mercedes_benz': 131068, 'ford': 125283, 'audi': 129126, 'opel': 129832}
Brand Mean Price Series Constructor audi             9363
bmw              8577
ford             3751
mercedes_benz    8495
opel             5447
volkswagen       6516
dtype: int64
Brand Mean Mileage Series Constructor audi             129126
bmw              132743
ford             125283
mercedes_benz    131068
opel             129832
volkswagen       129153
dtype: int64
               Mean_Price  Mean_Mileage
audi                 9363        129126
bmw                  8577        132743
ford                 3751        125283
mercedes_benz        8495        131068
opel                 5447        129832
volkswagen           6516        129153


The mileage for all the vehicles were similar, approximately 130,000km. The car prices ranged from 3751 USD for a Ford to 9363 USD for an Audi. The price for an Audi was nearly 2.5 times that of a Ford.